syft/execution/role.py

```diff
@@ -54,22 +50,47 @@ def input_placeholders(self):
     def output_placeholders(self):
         return [self.placeholders[id_] for id_ in self.output_placeholder_ids]
 
+    @staticmethod
+    def nested_object_traversal(obj, leaf_function, leaf_type):
+        if isinstance(obj, (list, tuple)):
+            result = [Role.nested_object_traversal(elem, leaf_function, leaf_type) for elem in obj]
+            return type(obj)(result)
+        elif isinstance(obj, dict):
+            return {
+                k: Role.nested_object_traversal(v, leaf_function, leaf_type)
+                for k, v in sorted(obj.items())
+            }
+        elif isinstance(obj, leaf_type):
+            return leaf_function(obj)
+        else:
+            return obj
+
     def register_inputs(self, args_):
         """ Takes input arguments for this role and generate placeholders.
         """
         # TODO Should we be able to rebuild?
-        self.input_placeholder_ids = tuple(
-            self._store_placeholders(arg).value for arg in args_ if isinstance(arg, PlaceHolder)
+
+        traversal_function = lambda x: self.input_placeholder_ids.append(
+            self._store_placeholders(x).value
         )
 
+        self.input_placeholder_ids = []
+        Role.nested_object_traversal(args_, traversal_function, PlaceHolder)
```

Comment:
Can't we use the return value of this function to fill `self.input_placeholder_ids`?

Reply:
`nested_object_traversal` returns the actuall "tree" scheme, while I want the flattened version of the tree, this would require a second pass through it, I don't think it's worth it.